### ISSUE TITLE
Wrong Tensor initialization with LongStorage causes Segmentation fault

### ISSUE THREAD

**Issue by dariosena** [torch/torch7#384]
The following **wrong** code causes segmentation fault and close the torch environment:
`l = torch.Storage(10):fill(1)`
`x = torch.Tensor(l, 1, torch.LongStorage(2,5))`
If I use another tensor type, instead of LongStorage, I got correctly `bad argument #3 to ...`

**Comment by soumith** [comment]
@dariosena torch.LongStorage's constructor's second argument is a pointer. Maybe, what you meant to do is torch.LongStorage(2 * 5).

Reference:
https://github.com/torch/torch7/blob/master/generic/Storage.c#L69-L71

**Comment by dariosena** [comment]
I agree! What I am trying to explain is the fact that if write the wrong code `x = torch.Tensor(l, 1, torch.FloatStorage(2,5))`, the torch environment is fault tolerant:
`bad argument #3 to '?' (number expected, got userdata)
stack traceback:
	[C]: at 0x7f85abe9c050
	[C]: in function 'Tensor'
	[string "x = torch.Tensor(l, 1, torch.FloatStorage(2,5))"]:1: in main chunk
	[C]: in function 'xpcall'
	/home/dario/torch/install/share/lua/5.1/trepl/init.lua:648: in function 'repl'
	...ario/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
	[C]: at 0x00406670	`
 
But with LongStorage I got `Segmentation fault (core dumped)` and the program returns to bash.

Thanks!

Emanuel D R Sena

**Comment by hughperkins** [comment]
Hi Emanuel,

Segfaults are a bit violent.  Is this something that you might have a moment to look into?  The call to `_(new)` arrives here https://github.com/torch/torch7/blob/master/generic/Tensor.c#L81 . Then, it checks the type of the first argument here https://github.com/torch/torch7/blob/master/generic/Tensor.c#L87 , and since it's not a table, it jumps to https://github.com/torch/torch7/blob/master/generic/Tensor.c#L198   You can see that there is not much argument checking at this point... but there are a few possible argument combinations, so might take a bid of fiddling/tweaking to get them all covered and tested.

**Comment by albanD** [comment]
There are two things here:

1) In the `Tensor` constructor, you try to use two different versions:
* When you give a `LongTensor` as third argument you use [this one](https://github.com/torch/torch7/blob/master/doc/tensor.md#torchtensorstorage-storageoffset-sizes-strides). In this case, the content of the tensor you give is the size of each dimension.
* When you give a `FloatTensor` it tries to use [this one](https://github.com/torch/torch7/blob/master/doc/tensor.md#torchtensorstorage-storageoffset-sz1--st1---sz4--st4) where it detects that the third argument is not a number but a tensor, hence the error message.

I think the error messages in the `Tensor` constructor are correct.

2) The wrong code that actually causes the segfault `torch.LongStorage(2,5)` is just wong because you create a storage pointing at address 5. The `Tensor` constructor segfault because it tries to access the storage containing the sizes. FYI if you do `torch.FloatStorage(2,5)` and try to access it, you will get a segfault as well.

@dariosena Do you still think this is an issue? I am not sure if we can actually verify that the second number you pass to the `TypeStorage` constructor is valid pointer or a number with another meaning (due to a typo) ?
